File: sdk/java/src/main/java/dev/udb/client/UdbPolicyBundleSignatureException.java
package dev.udb.client;

import com.udb.core.authz.services.v1.SignedPolicyBundle;

/**
 * Thrown by {@link UdbAuthClient#verifyPolicyBundle(SignedPolicyBundle, String)}
 * (and by {@link UdbAuthClient#getPolicyBundle()} when a bundle secret is set)
 * when the HMAC-SHA256 signature recomputed over the {@code bundle} bytes does
 * not match the server-supplied {@code signature}. Carries the offending bundle
 * so the caller can inspect its key id / policy version.
 */
public final class UdbPolicyBundleSignatureException extends RuntimeException {
  private final transient SignedPolicyBundle bundle;

  public UdbPolicyBundleSignatureException(String message, SignedPolicyBundle bundle) {
    super(message);
    this.bundle = bundle;
  }

  /** The bundle whose signature failed verification. */
  public SignedPolicyBundle bundle() {
    return bundle;
  }
}

File: sdk/csharp/Udb.Client/UdbPolicyBundleSignatureException.cs
namespace Udb.Client;

/// <summary>
/// Thrown by <see cref="UdbAuthClient.VerifyPolicyBundle"/> (and by
/// <see cref="UdbAuthClient.GetPolicyBundleAsync"/> when a bundle secret is
/// configured) when a signed policy bundle's HMAC-SHA256 signature does not
/// match the recomputed value — i.e. the bundle was tampered with or signed
/// under a different secret. Parity with the Go/Python/TS bundle-verification
/// error types (M5.3).
/// </summary>
public sealed class UdbPolicyBundleSignatureException : Exception
{
    public UdbPolicyBundleSignatureException(string keyId, string algorithm)
        : base($"udb: policy bundle signature verification failed (key_id='{keyId}', algorithm='{algorithm}')")
    {
        KeyId = keyId;
        Algorithm = algorithm;
    }

    /// <summary>The signing key id reported by the server, if any.</summary>
    public string KeyId { get; }

    /// <summary>The signature algorithm reported by the server (e.g. "HMAC-SHA256").</summary>
    public string Algorithm { get; }
}
